ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ExternalTableCopyTaskBuilder.java

```diff
@@ -0,0 +1,150 @@
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class ExternalTableCopyTaskBuilder {
+  private static final Logger LOG = LoggerFactory.getLogger(ExternalTableCopyTaskBuilder.class);
+  private final ReplLoadWork work;
+  private final HiveConf conf;
+
+  ExternalTableCopyTaskBuilder(ReplLoadWork work, HiveConf conf) {
+    this.work = work;
+    this.conf = conf;
+  }
+
+  List<Task<? extends Serializable>> tasks(TaskTracker tracker) {
+    List<Task<? extends Serializable>> tasks = new ArrayList<>();
+    Iterator<DirCopyWork> itr = work.getPathsToCopyIterator();
+    while (tracker.canAddMoreTasks() && itr.hasNext()) {
+      DirCopyWork dirCopyWork = itr.next();
+      Task<DirCopyWork> task = TaskFactory.get(dirCopyWork, conf);
+      tasks.add(task);
+      tracker.addTask(task);
+      LOG.debug("added task for {}", dirCopyWork);
+    }
+    return tasks;
+  }
+
+  public static class DirCopyTask extends Task<DirCopyWork> implements Serializable {
+    private static final Logger LOG = LoggerFactory.getLogger(DirCopyTask.class);
+    private static final int MAX_COPY_RETRY = 5;
+
+    @Override
+    protected int execute(DriverContext driverContext) {
+      String distCpDoAsUser = conf.getVar(HiveConf.ConfVars.HIVE_DISTCP_DOAS_USER);
+
+      Path sourcePath = work.fullyQualifiedSourcePath;
+      Path targetPath = work.fullyQualifiedTargetPath;
+      if (conf.getBoolVar(HiveConf.ConfVars.REPL_ADD_RAW_RESERVED_NAMESPACE)) {
+        sourcePath = reservedRawPath(work.fullyQualifiedSourcePath.toUri());
+        targetPath = reservedRawPath(work.fullyQualifiedTargetPath.toUri());
+      }
+      int currentRetry = 0;
+      while (currentRetry < MAX_COPY_RETRY) {
+        try {
+          UserGroupInformation ugi = Utils.getUGI();
+          String currentUser = ugi.getShortUserName();
+          boolean usePrivilegedUser =
+              distCpDoAsUser != null && !currentUser.equals(distCpDoAsUser);
+
+          // do we create a new conf and only here provide this additional option so that we get away from
+          // differences of data in two location for the same directories ?
+          // basically add distcp.options.delete to hiveconf new object ?
+          FileUtils.distCp(
+              sourcePath.getFileSystem(conf), // source file system
+              Collections.singletonList(sourcePath),  // list of source paths
+              targetPath,
+              false,
+              usePrivilegedUser ? distCpDoAsUser : null,
+              conf,
+              ShimLoader.getHadoopShims());
+          return 0;
+        } catch (Exception e) {
+          if (++currentRetry < MAX_COPY_RETRY) {
+            LOG.warn("unable to copy", e);
+          } else {
+            LOG.error("unable to copy {} to {}", sourcePath, targetPath, e);
+            setException(e);
+            return ErrorMsg.getErrorMsg(e.getMessage()).getErrorCode();
```

Comment:
well we wont know that, The idea behind external table is that the actual data is not deleted on drop table. so they should not delete the location on the source till its replicated. we cant also skip this since if there is a bug or something in our code which is looking at wrong path then  we will just keep skipping the copy of data.

Reply:
make sense. One more question. Who is responsible to delete the external table location in target if the table is dropped?

Reply:
the user for now.